File: test_MLMC_runs/check_variances.py
import time
import subprocess
import numpy as np
import os
os.system('rm -fv log.check_variances')

def sample_cpfem(level):
    cmd = ["python3", "wrapper_DREAM3D-DAMASK.py", "--level=" + str(level)]
    out = subprocess.run(cmd, stdout=subprocess.PIPE).stdout.decode("utf-8") # execute command and read output
    lines = out.splitlines() # split output into lines
    Q = [] # empty array
    for line in lines: # loop over all lines
        if "Estimated Yield Stress" in line: # if line containes "Estimated Yield Stress"
            try:
                yield_stress = float(line.split()[-2]) # parse the yield stress value to Float64
                Q.append(yield_stress) # append the value of the yield stress to the array "Q"
            except:
                Q.append(None)
    print(Q)
    if len(Q) == 2: # safeguard
        if level == 0:
            return Q[0], Q[0] # return twice Qf
        else:
            return Q[0] - Q[1], Q[0] # return multilevel difference and Qf
    else:
        return None, None # safeguard

def check_variances(max_level=3, budget=3600*24*2):
    buget_per_level = budget/(max_level + 1)

    # for level in range(max_level):
    for level in range(max_level - 1, -1, -1): # reverse order
        samps_dQ = []
        samps_Qf = []
        timer = 0
        while timer < buget_per_level:
            t = time.time()
            dQ, Qf = sample_cpfem(level)
            if dQ != None and Qf != None: # if both dQ and Qf are not both None
                timer += time.time() - t
                samps_dQ.append(dQ)
                samps_Qf.append(Qf)
                print("level =", level,
                        ", E =", np.mean(samps_Qf),
                        ", dE =", np.mean(samps_dQ),
                        ", V =", np.var(samps_Qf),
                        ", dV =", np.var(samps_dQ),
                        ",", len(samps_dQ), "samples, time per sample is", timer/len(samps_dQ))
                logFile = open('log.check_variances', 'a')
                logFile.write("level = %d, E = %.8f, dE = %.8f, V = %.8f, dV = %.8f, len(samps_dQ) = %d samples, time per sample is %.8f\n" % (level, np.mean(samps_Qf), np.mean(samps_dQ), np.var(samps_Qf), np.var(samps_dQ), len(samps_dQ), timer/len(samps_dQ)))


if __name__ == "__main__":
    check_variances()

File: test_MLMC_runs/check_variances.jl
###############################################################################
#                                                                             #
#                            check_variances.jl                               #
#                                                                             #
###############################################################################

using MultilevelEstimators
using Statistics
using PrettyTables
using ProgressMeter

# Command to run multi-index version of DREAM3D-DAMASK
get_cmd(index::Index) = Cmd(["python3", "wrapperMLMC_DREAM3D-DAMASK.py", "-level", string(index)])

# Read DREAM3D-DAMASK wrapper output and return value of estimated yield
# stress for the given level or index
function get_qoi(out, index)
    for line in split(out, "\n")
        if occursin("Estimated Yield Stress at $(index)", line)
            return parse(Float64, split(line)[end - 1])
        end
    end
    return NaN # return NaN if no value could be found
end

# Compute a sample of the multilevel or multi-index difference at the given
# level or index
function sample(index::Index)
    cmd = get_cmd(index)
    out = read(cmd, String)
    Qf = get_qoi(out, index)
    dQ = Qf
    for (key, val) in diff(index)
        Qc = get_qoi(out, key)
        dQ += val * Qc
    end
    return dQ, Qf
end

# Default color in table header
get_crayon() = crayon"yellow bold"

# Column highlighter
highlight_cols(k) = Highlighter((v, i, j) -> any(j .== k), get_crayon())

# Print summary statistics
function print_table(data)
    ell_name = length(first(keys(data))) > 1 ? "Index" : "Level"
    header = [ell_name, "|Eℓ|", "|ΔEℓ|", "Vℓ", "ΔVℓ", "Nℓ", "Wℓ"]
    E = [abs(mean(val[2])) for (key, val) in data]
    dE = [abs(mean(val[1])) for (key, val) in data]
    V = [var(val[2]) for (key, val) in data]
    dV = [var(val[1]) for (key, val) in data]
    N = [length(val[1]) for (key, val) in data]
    W = [length(val[1])/val[3] for (key, val) in data]
    entries = hcat(sort(collect(keys(data))), E, dE, V, dV, N, W)
    e_fmt = ft_printf("%5.3e", [2, 3, 4, 5, 7])
    pretty_table(entries,
                 header,
                 header_crayon=get_crayon(),
                 header_alignment=:c, 
                 formatters = e_fmt,
                 highlighters=highlight_cols(1),
                 equal_columns_width=true)
end

"""
check_variances([index_set=ML()], [max_level=3], [budget=20])

Check multilevel or multi-index variance decay for level or index set
`index_set` with maximum level parameter `max_level` by taking samples 
on each index for approximately `buget` seconds.
"""
function check_variances(; index_set=ML(), max_level=3, budget=20)
    indices = collect(get_index_set(index_set, max_level))
    nb_of_indices = length(indices)
    budget_per_level = round(Int, budget/(nb_of_indices + 1)) # split budget evenly over all levels/indices
    data = Dict{Index, Any}() # empty dictionary to hold samples

    # Loop over all levels/indices
    for index in indices
        samples_Qf = [] # vector to store samples of Qf
        samples_dQ = [] # vector to store samples of dQ
        timer = 0
        p = Progress(budget_per_level, dt=1, barglyphs=BarGlyphs("[=> ]"), color=:none)
        while timer < budget_per_level # run timer until we run out of buget
            timer += @elapsed dQ, Qf = sample(index) # take a new sample
            if isfinite(dQ) && isfinite(Qf) # check if the sample is valid
                push!(samples_dQ, dQ) # update dQ
                push!(samples_Qf, Qf) # update Qf
            end
            update!(p, min(budget_per_level, floor(Int, timer))) # update progress bar
        end
        data[index] = (samples_dQ, samples_Qf, timer) # add new key to the dictionary
        print_table(data) # print the results
    end
end
